src/ethereum/prague/vm/eoa_delegation.py

```diff
@@ -0,0 +1,182 @@
+"""
+Set EOA account code.
+"""
+
+
+from typing import Tuple
+
+from ethereum import rlp
+from ethereum.base_types import U64, U256, Bytes, Uint
+from ethereum.crypto.elliptic_curve import SECP256K1N, secp256k1_recover
+from ethereum.crypto.hash import keccak256
+from ethereum.exceptions import InvalidBlock
+
+from ..fork_types import Address, Authorization
+from ..state import account_exists, get_account, increment_nonce, set_code
+from ..vm.gas import GAS_COLD_ACCOUNT_ACCESS, GAS_WARM_ACCESS
+from . import Environment, Evm, Message
+from .exceptions import InvalidAuthorization
+
+SET_CODE_TX_MAGIC = b"\x05"
+EOA_DELEGATION_MARKER = b"\xEF\x01\x00"
+EOA_DELEGATION_MARKER_LENGTH = len(EOA_DELEGATION_MARKER)
+EOA_DELEGATED_CODE_LENGTH = 23
+PER_EMPTY_ACCOUNT_COST = 25000
+PER_AUTH_BASE_COST = 2500
+
+
+def is_valid_delegation(code: bytes) -> bool:
+    """
+    Whether the code is a valid delegation designation.
+
+    Parameters
+    ----------
+    code: `bytes`
+        The code to check.
+
+    Returns
+    -------
+    valid : `bool`
+        True if the code is a valid delegation designation,
+        False otherwise.
+    """
+    if (
+        len(code) == EOA_DELEGATED_CODE_LENGTH
+        and code[:EOA_DELEGATION_MARKER_LENGTH] == EOA_DELEGATION_MARKER
+    ):
+        return True
+    return False
+
+
+def recover_authority(authorization: Authorization) -> Address:
+    """
+    Recover the authority address from the authorization.
+
+    Parameters
+    ----------
+    authorization
+        The authorization to recover the authority from.
+
+    Returns
+    -------
+    authority : `Address`
+        The recovered authority address.
+    """
+    y_parity, r, s = authorization.y_parity, authorization.r, authorization.s
+    if 0 >= r or r >= SECP256K1N:
+        raise InvalidAuthorization
+    if 0 >= s or s > SECP256K1N // 2:
+        raise InvalidAuthorization
```

Comment:
As it currently stands failing the check on `s` invalidates the entire transaction. It should be in `check_transaction()`. Surprisingly, the same doesn't apply to the check on `r` (this is probably an oversight by the EIP author).

There is [a PR](https://github.com/ethereum/EIPs/pull/8845) in the EIPS repo to change the behaviour to what you have implemented, but it is not merged.

Reply:
ACDE decided to adopt the aforementioned PR so this code is now correct!